engine/physics/src/physics/box2d/box2d_physics.cpp

```diff
@@ -0,0 +1,2555 @@
+                    b2DestroyShape(shape->m_ShapeId, false);
+                PolygonShapeData* poly_shape = (PolygonShapeData*) shape;
+                delete poly_shape;
+            }
+            break;
+            case SHAPE_TYPE_GRID:
+            {
+                GridShapeData* grid_shape = (GridShapeData*) shape;
+
+                uint32_t cell_count = grid_shape->m_RowCount * grid_shape->m_ColumnCount;
+
+                for (int i = 0; i < cell_count; ++i)
+                {
+                    if (grid_shape->m_Cells[i].m_Index != B2GRIDSHAPE_EMPTY_CELL)
+                    {
+                        if (b2Shape_IsValid(grid_shape->m_CellPolygonShapes[i]))
+                            b2DestroyShape(grid_shape->m_CellPolygonShapes[i], false);
+                    }
+                }
+                delete grid_shape;
+            } break;
+            default: break;
+        }
+    }
+
+    /*
+     * NOTE: In order to support shape transform we create a copy of shapes using the function TransformCopyShape() above
+     * This is required as the transform is part of the shape and due to absence of a compound shape, aka list shape
+     * with per-child transform.
+     */
+    HCollisionObject2D NewCollisionObject2D(HWorld2D world, const CollisionObjectData& data, HCollisionShape2D* shapes,
+                                            Vector3* translations, Quat* rotations,
+                                            uint32_t shape_count)
+    {
+        if (shape_count == 0)
+        {
+            dmLogError("Collision objects must have a shape.");
+            return 0x0;
+        }
+        switch (data.m_Type)
+        {
+        case COLLISION_OBJECT_TYPE_DYNAMIC:
+            if (data.m_Mass == 0.0f)
+            {
+                dmLogError("Collision objects can not be dynamic and have zero mass.");
+                return 0x0;
+            }
+            break;
+        default:
+            if (data.m_Mass > 0.0f)
+            {
+                dmLogError("Only dynamic collision objects can have a positive mass.");
+                return 0x0;
+            }
+            break;
+        }
+
+        HContext2D context = world->m_Context;
+        b2BodyDef def = b2DefaultBodyDef();
+
+        float scale = 1.0f;
+        if (world->m_GetWorldTransformCallback != 0x0)
+        {
+            if (data.m_UserData != 0x0)
+            {
+                dmTransform::Transform world_transform;
+                (*world->m_GetWorldTransformCallback)(data.m_UserData, world_transform);
+
+                Point3 position = Point3(world_transform.GetTranslation());
+
+                // I don't think this is the way to go, but tests return broken transforms..
+                position.setX(dmMath::Clamp(position.getX(), -1000.0f, 1000.0f));
+                position.setY(dmMath::Clamp(position.getY(), -1000.0f, 1000.0f));
+                position.setZ(dmMath::Clamp(position.getZ(), -1000.0f, 1000.0f));
```

Comment:
If I remove these clamps, tests will unfortunately fail due to the transform we get back contains weird / broken data. The old implementation of box2d didn't seem to care, but the new version does. I'm not sure if it's possible to check the transform for validity or not?